clients/venice-push-job/src/test/java/com/linkedin/venice/spark/input/pubsub/raw/VeniceRawPubsubInputPartitionReaderTest.java

```diff
@@ -0,0 +1,302 @@
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.testng.Assert;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+
+/**
+ * Unit tests for {@link VeniceRawPubsubInputPartitionReader}.
+ *
+ * This test class validates the functionality of the partition reader including:
+ * - Reading messages from PubSub topic partition
+ * - Handling of starting and ending offsets
+ * - Conversion of PubSub messages to Spark InternalRow format
+ * - Filtering of control messages
+ * - Progress tracking and reporting
+ * - Error conditions and edge cases
+ */
+public class VeniceRawPubsubInputPartitionReaderTest {
+  @Mock
+  private VeniceBasicPubsubInputPartition mockInputPartition;
+
+  @Mock
+  private PubSubConsumerAdapter mockConsumer;
+
+  @Mock
+  private PubSubTopic mockTopic;
+
+  @Mock
+  private PubSubTopicPartition mockTopicPartition;
+
+  @Mock
+  private VenicePubSubMessageToRow mockMessageToRowConverter;
+
+  private VeniceRawPubsubInputPartitionReader reader;
+
+  @BeforeMethod
+  public void setUp() {
+    MockitoAnnotations.openMocks(this);
+    // Setup will be implemented
+  }
+
+  @AfterMethod
+  public void tearDown() {
+    if (reader != null) {
+      reader.close();
+    }
+  }
+
+  @Test
+  public void testEmptyTopicInitialization() {
+    // Arrange
+    int partitionNumber = 5;
+    String topicName = "test-topic";
+    String region = "test-region";
+    long startOffset = 101L;
+    long endOffset = 200L;
+
+    // updated constants to adjust the test behavior and timeouts
+    final int CONSUMER_POLL_EMPTY_RESULT_RETRY_TIMES = 3;
+    final long EMPTY_POLL_SLEEP_TIME_MS = TimeUnit.SECONDS.toMillis(1);
+    final long CONSUMER_POLL_TIMEOUT = TimeUnit.SECONDS.toMillis(1); // 1 second
+
+    // Setup mocks
+    when(mockInputPartition.getPartitionNumber()).thenReturn(partitionNumber);
+    when(mockInputPartition.getTopicName()).thenReturn(topicName);
+    when(mockInputPartition.getRegion()).thenReturn(region);
+    when(mockInputPartition.getStartOffset()).thenReturn(startOffset);
+    when(mockInputPartition.getEndOffset()).thenReturn(endOffset);
+    when(mockTopicPartition.getPubSubTopic()).thenReturn(mockTopic);
+
+    // Act
+    long start = System.currentTimeMillis();
+    reader = new VeniceRawPubsubInputPartitionReader(
+        mockInputPartition,
+        mockConsumer,
+        mockTopicPartition,
+        CONSUMER_POLL_TIMEOUT,
+        CONSUMER_POLL_EMPTY_RESULT_RETRY_TIMES,
+        EMPTY_POLL_SLEEP_TIME_MS,
+        new VenicePubSubMessageToRow());
+
+    Assert.assertFalse(reader.next()); // Attempt to read messages, which should fail due to empty topic
+
+    long elapsed = System.currentTimeMillis() - start;
+    Assert.assertTrue(elapsed >= 3000, "Constructor should take at least 3 seconds due to polling retries.");
```

Comment:
Nit, the value here is related to CONSUMER_POLL_EMPTY_RESULT_RETRY_TIMES and other constants there, we could just use that to improve readablity.